README.md
# Specify: A dimensioning script for Adobe Illustrator

Adobe Illustrator script and extension to automate specifying dimension measurements (and adding dimension lines) of single or multiple objects, and specifying the dimensions between two objects.

![specify-example](ext/img/featured.png)

## Installation

### Extension Install

Simply [download the Specify extension from Adobe Exchange](https://exchange.adobe.com/creativecloud.details.104123.specify.html).

### Script Install

The instructions below only pertain to the script-only (non-extension) version of Specify.

1. [Download a copy of the latest release](https://github.com/adamdehaven/Specify/archive/master.zip).
2. Extract the contents of the `.zip` file to one of the recommended file paths shown below (depending on your OS) in order to allow running the script from Illustrator's `File > Scripts` menu.

    ``` sh
    # Windows
    C:\Program Files\Adobe\Adobe Illustrator [VERSION]\Presets\en_US\Scripts

    # Mac
    Applications > Adobe Illustrator [VERSION] > Presets > en_US > Scripts
    ```

Alternatively, you may save the folder and all of its contents at the location of your choosing.

Regardless of where you save the files, you **must** save the entire `Specify` folder and all of its contents at the same location, without manipulating the structure of the folder.

## Usage

To run the extension or the script, first select one or more objects in your open Illustrator document. You must have at least one object selected via Illustrator's Selection Tool <kbd>V</kbd> or Direct Selection Tool <kbd>A</kbd>.

### Extension Usage

After [installing the Specify extension from Adobe Exchange](https://exchange.adobe.com/creativecloud.details.104123.specify.html), simply navigate to `Window > Extensions > Specify` to get started. Run the script by simply clicking the "Specify Object(s)" button in the extension panel.

### Script Usage

Assuming you saved the script as [described above](#installation), in Illustrator's menu, go to `File > Scripts > Specify > Specify` ("Specify" appears twice because you have to call the script file inside the Specify directory).

If you chose an alternate location to save the Specify folder, you can run by going to `File > Scripts > Other Script...`, and then selecting the `Specify.jsx` file.

### Options and Styles

When the script initializes, you will be presented with a dialog that allows you to choose which dimension(s) of your object(s) you would like to specify. The dialog also allows for the customization of several options. Any changes to the settings defined in the Specify Dialog (including both the [Options Panel](#options-panel) and the [Styles Panel](#styles-panel)) will persist until you close the Adobe Illustrator application, even if you run the script in multiple documents.

After configuring your desired settings, make sure to select which dimension(s) to dimension, and then simply click the "Specify Object(s)" button at the bottom right of the dialog. Specify will quickly output all selected dimensions onto a new **SPEC** layer, and then lock the layer to prevent accidental edits.

### SPEC Layer

You are free to unlock the **SPEC** layer and edit as needed. You may change the color of individual dimension labels, adjust the placement of dimension groups, or even remove dimensions you no longer need.

If you make a mistake, no problem! Simply select the desired objects and run the script again.

## Options Panel

The Options Panel allows the user to select the side(s) of the object(s) to dimension, whether to dimension the object itself or between objects, and the scale of the artwork.

![Specify Illustrator Script Options Panel](img/specify-options-panel.png)

### Dimensions

The options panel allows you to first choose which dimension(s) of your Illustrator object you would like to specify. You may choose top, right, bottom, left, or any combination of sides.

### Multiple Objects

If exactly two objects are selected on your artboard, you will be given the option to dimension between the two objects, instead of their individual sides. When checked, this option will specify the selected dimension(s) between the two chosen objects.

### Scale

You may define a custom scale for your artwork, which will apply a multiplier to the displayed units label.

For example, if you are designing an engineering drawing at 1/4 scale (or 1:4), you simply choose this ratio from the **Scale** dropdown menu, and then optionally enable the checkbox for [Custom Units Text](#custom-units-text) on the Styles Panel that corresponds to the unit labels you would like to output. In this example, choosing '1/4' will indicate the artwork is drawn at one-fourth scale, resulting in dimension values that are 4 times their drawn dimensions.

## Styles Panel

The Styles Panel allows the user to completely customize the label styles and the line styles for all dimensions.

![Specify Illustrator Script Styles Panel](img/specify-styles-panel.png)

### Include Units in Label

When checked, inserts the units in the label alongside the dimension measurement.

The script automatically dimensions objects based on the Document's default units. To dimension in different units (i.e. Inches, Centimeters, etc.), first, select show Rulers via `View > Rulers > Show Rulers` ( <kbd>⌘Cmd</kbd> + <kbd>R</kbd> on Mac, <kbd>Ctrl</kbd> + <kbd>R</kbd> on PC ).

Next, right-click on the Ruler, and select your desired units. Otherwise, the script will use the Document's selected units by default.

### Custom Units Text

When checked, allows the user to customize the text of the units label. For example, displaying the abbreviation for feet, "ft" instead of the ruler measurement unit.

### Decimals

Set the desired number of decimal places to display after the decimal point in the label dimensions.

### Font Size

Set the desired font size for the dimension label(s). If the value is less than one (e.g. 0.25) you must include a leading zero before the decimal point.

### Color

Click to select the color for the dimension label(s).

### Gap

Set the size of the desired gap between the dimension label(s) and the object.

### Stroke Width

Set the desired stroke width of the dimension line.

### Head & Tail Length

Set the desired length of the line at both ends of the dimension line.

## Updates Panel

The Updates Panel provides the current script version information, along with references to the GitHub code repository, and author information.

![Specify Illustrator Script Updates Panel](img/specify-updates-panel.png)

Be sure to star and subscribe [on GitHub](https://github.com/adamdehaven/specify) in order to stay up-to-date on future releases.

## Reset

At the bottom left of the Specify dialog is a Reset button that restores all defaults in both the Options Panel and the Styles panel.

## Support

Do you have questions? Would you like to contribute? Something else? Feel free to [open an issue on GitHub](https://github.com/adamdehaven/Specify/issues).

Not into looking at code? No worries. Reach out on Twitter [@adamdehaven](https://twitter.com/adamdehaven).

## Examples

![specify-example](img/specify-example.jpg)

![specify-workplace-example](img/specify-workplace-example.png)


CHANGELOG.md
# Specify Changelog

## 2.0.4

### Fixes

- Adjusting default stroke with to 1 unit (instead of 0.5)
- Updated internal build process

## 2.0.2

### New Features

- Completely updated the Specify dialog UI
- **New Dialog UI and Adobe Extension**: Added a version packaged as a [Free Creative Cloud Extension available via Adobe Exchange](https://exchange.adobe.com/creativecloud.details.104123.specify.html) for Illustrator CC users.

## 1.3.1

### Fixes

- Automatically adds a leading zero to the font size field if user enters a value such as `.25` instead of `0.25`. For reference, font sizes require a leading zero on input.

## 1.3.0

### New Features

- **User-defined custom scale for dimensions and labels**: Allows the user to specify a working scale along with allowing for customizable unit labels. This release enables the user to specify a working scale along with allowing for customizable unit labels.

## 1.2.2

### Fixes

- Prevents the script from generating a dashed line when specifying the dimensions of an object with dashed paths.

## 1.2.1

### Fixes

- The font size input in the options panel now allows the user to specify the size of the label text in the same units as the document ruler. 
- Fixes a bug with the help text that appears when hovering over an input field in the Specify dialog box. Help text was showing the current settings as Defaults rather than the script's default values.

## 1.2.0

### New Features

- Implemented environmental variables to persist settings (to persist options between running the script multiple times, even between multiple documents) until application is closed.
- Added a 'Restore Defaults' button to roll back to initial settings.

### Fixes

- Removed 'Advanced Options' toggle and section; moved all items into 'Options' panel.

## 1.1.0

### New Features

- Now possible to specify multiple objects simultaneously. 
- Added 'Select All' option to easily select all dimensions to specify.
- Added 'Advanced Options' panel:
  - Easily change font size.
  - Easily change RGB color of dimension lines and labels.
  - Easily change number of decimal places for dimension labels.
- Added help tips (hover over options in dialog).

### Fixes

- Updated 'Specify' button and added 'Cancel' button.
- Improved error handling.
- Fixed issue with label spacing.
- Improved UI of dialog box and options.

## 1.0.0

Initial release.


index.js
function onLoaded() {
    // Create an instance of CSInterface
    var csInterface = new CSInterface();

    // Set the initial panel colors
    updateThemeWithAppSkinInfo(csInterface.hostEnvironment.appSkinInfo);

    // Bind event for extension panel window visiblilty change
    csInterface.addEventListener("com.adobe.csxs.events.WindowVisibilityChanged", function (e) {
        var specifyButton = document.getElementById('specify-button');
        if (specifyButton) {
            // Re-enable specifyButton
            specifyButton.disabled = false;
        }
        // Update the color of the panel when the theme color of the product changed.
        onAppThemeColorChanged();
    });

    function runSpecifyScript(e) {
        e.preventDefault();
        if (specifyButton.disabled == true) {
            return
        }

        csInterface.evalScript("specifyObjects(true)");
    }

    var specifyButton = document.getElementById('specify-button');
    if (specifyButton) {
        // Run script on click if button is not disabled
        specifyButton.addEventListener("click", runSpecifyScript);
    }

    // Respond to custom event of dialog toggled open or close
    csInterface.addEventListener("com.adamdehaven.specify.dialogToggled", function (e) {
        // Enable/disable extension panel button
        if (specifyButton) {
            if (e.data === 'open') {
                specifyButton.disabled = true;
            } else if (e.data === 'close') {
                specifyButton.disabled = false;
            }
        }

    });

}
/**
 * Update the theme with the AppSkinInfo retrieved from the host product.
 */
function updateThemeWithAppSkinInfo(appSkinInfo) {

    //Update the background color of the panel
    var panelBackgroundColor = appSkinInfo.panelBackgroundColor.color;
    document.body.bgColor = toHex(panelBackgroundColor);

    var styleId = "specify-extension-styles";

    var csInterface = new CSInterface();

    var isPanelThemeLight = panelBackgroundColor.red > 127;
    var buttonColor;
    var buttonBorderColor;
    var buttonBackgroundColor;

    if (isPanelThemeLight) {
        buttonColor = "#000000 !important;";
        buttonBorderColor = toHex(panelBackgroundColor, -90) + " !important;";
        buttonBackgroundColor = toHex(panelBackgroundColor, 54) + " !important;";
    } else {
        buttonColor = "#ffffff !important;";
        buttonBorderColor = toHex(panelBackgroundColor, -45) + " !important;";
        buttonBackgroundColor = toHex(panelBackgroundColor, -20) + " !important;";
    }

    addRule(styleId, ".specify-extension-panel, .default", "font-size:" + appSkinInfo.baseFontSize + "px" + "; color:" + reverseColor(panelBackgroundColor) + "; background-color:" + toHex(panelBackgroundColor, 1));

    addRule(styleId, "#specify-button", "border-color: " + buttonBorderColor);
    addRule(styleId, "#specify-button", "background-color: " + buttonBackgroundColor);
    addRule(styleId, "#specify-button", "color: " + buttonColor);

}

function addRule(stylesheetId, selector, rule) {
    var stylesheet = document.getElementById(stylesheetId);

    if (!stylesheet) {
        stylesheet = document.createElement('style');
        stylesheet.id = stylesheetId;
        document.head.appendChild(stylesheet)
    }

    try {
        stylesheet = stylesheet.sheet;
        if (stylesheet.addRule) {
            stylesheet.addRule(selector, rule);
        } else if (stylesheet.insertRule) {
            stylesheet.insertRule(selector + ' { ' + rule + ' }', stylesheet.cssRules.length);
        }
    } catch (e) {

    }
}

function reverseColor(color, delta) {
    return toHex({ red: Math.abs(255 - color.red), green: Math.abs(255 - color.green), blue: Math.abs(255 - color.blue) }, delta);
}

/**
 * Convert the Color object to string in hexadecimal format;
 */
function toHex(color, delta) {
    function computeValue(value, delta) {
        var computedValue = !isNaN(delta) ? value + delta : value;
        if (computedValue < 0) {
            computedValue = 0;
        } else if (computedValue > 255) {
            computedValue = 255;
        }

        computedValue = Math.round(computedValue).toString(16);
        return computedValue.length == 1 ? "0" + computedValue : computedValue;
    }

    var hex = "";
    if (color) {
        with (color) {
            hex = computeValue(red, delta) + computeValue(green, delta) + computeValue(blue, delta);
        };
    }
    return "#" + hex;
}

function onAppThemeColorChanged(event) {
    // Should get a latest HostEnvironment object from application.
    var skinInfo = JSON.parse(window.__adobe_cep__.getHostEnvironment()).appSkinInfo;
    // Gets the style information such as color info from the skinInfo,
    // and redraw all UI controls of your extension according to the style info.
    updateThemeWithAppSkinInfo(skinInfo);
}

extension/js/ext.js
function onLoaded() {
    // Create an instance of CSInterface
    var csInterface = new CSInterface();

    // Set the initial panel colors
    updateThemeWithAppSkinInfo(csInterface.hostEnvironment.appSkinInfo);

    // Bind event for extension panel window visiblilty change
    csInterface.addEventListener("com.adobe.csxs.events.WindowVisibilityChanged", function (e) {
        var specifyButton = document.getElementById('specify-button');
        if (specifyButton) {
            // Re-enable specifyButton
            specifyButton.disabled = false;
        }
        // Update the color of the panel when the theme color of the product changed.
        onAppThemeColorChanged();
    });

    function runSpecifyScript(e) {
        e.preventDefault();
        if (specifyButton.disabled == true) {
            return
        }

        csInterface.evalScript("specifyObjects(true)");
    }

    var specifyButton = document.getElementById('specify-button');
    if (specifyButton) {
        // Run script on click if button is not disabled
        specifyButton.addEventListener("click", runSpecifyScript);
    }

    // Respond to custom event of dialog toggled open or close
    csInterface.addEventListener("com.adamdehaven.specify.dialogToggled", function (e) {
        // Enable/disable extension panel button
        if (specifyButton) {
            if (e.data === 'open') {
                specifyButton.disabled = true;
            } else if (e.data === 'close') {
                specifyButton.disabled = false;
            }
        }
    });

    // Respond to external link click
    csInterface.addEventListener("com.adamdehaven.specify.openExternalLink", function (e) {
        // Get URL
        var url = e.data;
        if (url) {
            csInterface.openURLInDefaultBrowser(url);
        }
    });

}
/**
 * Update the theme with the AppSkinInfo retrieved from the host product.
 */
function updateThemeWithAppSkinInfo(appSkinInfo) {

    //Update the background color of the panel
    var panelBackgroundColor = appSkinInfo.panelBackgroundColor.color;
    document.body.bgColor = toHex(panelBackgroundColor);

    var styleId = "specify-extension-styles";

    var csInterface = new CSInterface();

    var isPanelThemeLight = panelBackgroundColor.red > 127;
    var buttonColor;
    var buttonBorderColor;
    var buttonBackgroundColor;

    if (isPanelThemeLight) {
        buttonColor = "#000000 !important;";
        buttonBorderColor = toHex(panelBackgroundColor, -90) + " !important;";
        buttonBackgroundColor = toHex(panelBackgroundColor, 54) + " !important;";
    } else {
        buttonColor = "#ffffff !important;";
        buttonBorderColor = toHex(panelBackgroundColor, -45) + " !important;";
        buttonBackgroundColor = toHex(panelBackgroundColor, -20) + " !important;";
    }

    addRule(styleId, ".specify-extension-panel, .default", "font-size:" + appSkinInfo.baseFontSize + "px" + "; color:" + reverseColor(panelBackgroundColor) + "; background-color:" + toHex(panelBackgroundColor, 1));

    addRule(styleId, "#specify-button", "border-color: " + buttonBorderColor);
    addRule(styleId, "#specify-button", "background-color: " + buttonBackgroundColor);
    addRule(styleId, "#specify-button", "color: " + buttonColor);

}

function addRule(stylesheetId, selector, rule) {
    var stylesheet = document.getElementById(stylesheetId);

    if (!stylesheet) {
        stylesheet = document.createElement('style');
        stylesheet.id = stylesheetId;
        document.head.appendChild(stylesheet)
    }

    try {
        stylesheet = stylesheet.sheet;
        if (stylesheet.addRule) {
            stylesheet.addRule(selector, rule);
        } else if (stylesheet.insertRule) {
            stylesheet.insertRule(selector + ' { ' + rule + ' }', stylesheet.cssRules.length);
        }
    } catch (e) {

    }
}

function reverseColor(color, delta) {
    return toHex({ red: Math.abs(255 - color.red), green: Math.abs(255 - color.green), blue: Math.abs(255 - color.blue) }, delta);
}

/**
 * Convert the Color object to string in hexadecimal format;
 */
function toHex(color, delta) {
    function computeValue(value, delta) {
        var computedValue = !isNaN(delta) ? value + delta : value;
        if (computedValue < 0) {
            computedValue = 0;
        } else if (computedValue > 255) {
            computedValue = 255;
        }

        computedValue = Math.round(computedValue).toString(16);
        return computedValue.length == 1 ? "0" + computedValue : computedValue;
    }

    var hex = "";
    if (color) {
        with (color) {
            hex = computeValue(red, delta) + computeValue(green, delta) + computeValue(blue, delta);
        };
    }
    return "#" + hex;
}

function onAppThemeColorChanged(event) {
    // Should get a latest HostEnvironment object from application.
    var skinInfo = JSON.parse(window.__adobe_cep__.getHostEnvironment()).appSkinInfo;
    // Gets the style information such as color info from the skinInfo,
    // and redraw all UI controls of your extension according to the style info.
    updateThemeWithAppSkinInfo(skinInfo);
}